tbot_contrib/gpio.py

```diff
@@ -24,20 +41,77 @@ def set_direction(self, direction: str) -> None:
         self._direction = direction
 
     def set_active_low(self, value: bool) -> None:
+        """Set Active Low
+
+        When Active Low is ``True``, the GPIO triggers on a ``LOW`` singnal, else on a ``HIGH`` signal
+
+        Args:
+            value (bool): Value
+        """
         (self._gpio_path / "active_low").write_text("1" if value else "0")
 
     def get_active_low(self) -> bool:
+        """Get Active Low
+
+        Reads value and returns direction ``in`` or ``out``
+
+        Returns:
+            bool: Value
+        """
         return (self._gpio_path / "active_low").read_text().strip() != "0"
 
     def get_direction(self) -> str:
+        """Get Direction of GPIO
+
+        Reads direction and returns direction ``in`` or ``out``
+
+        Returns:
+            str: Direction
+        """
         return (self._gpio_path / "direction").read_text().strip()
 
     def set_value(self, value: bool) -> None:
+        """Set Value
+
+        Sets value to ``HIGH`` and ``LOW``
+
+        Args:
+            value (bool): Value to set
+
+        Raises:
+            Exception: Direction
+        """
         if self._direction == "in":
             raise Exception("Can't set a GPIO which is not an output")
         (self._gpio_path / "value").write_text("1" if value else "0")
 
     def get_value(self) -> bool:
+        """Get Value
+
+        Reads value and returns value ``True`` or ``False``
+
+        Raises:
+            Exception: Direction
+
+        Returns:
+            bool: Value
+        """
         if not (self._direction == "in"):
             raise Exception("Can't get a value from a GPIO which is not an input")
         return (self._gpio_path / "value").read_text().strip() != "0"
+
+    def toggle_value(self, delay: float) -> None:
+        """Toggle GPIO
+
+        Toggles GPIO Pin between ``HIGH`` and ``LOW``
+
+        Args:
+            delay (float): Delay in s
+
+        Raises:
+            Exception: Direction
+        """
+        self.set_value(True)
+        time.sleep(delay)
+        self.set_value(False)
+        time.sleep(delay)
```

Comment:
I don't think _toggle_ is the right name for this.  To toggle a pin usually refers to set it LOW when it was previously HIGH and vice versa.  Not set and reset it in a single operation.

I would suggest calling it `pulse` and then have parameters `on_time` and `off_time` where `off_time` defaults to `None` and `None` meaning "use on_time for off_time".  In code:

```python
from typing import Optional

def pulse(self, on_time: float, off_time: Optional[float] = None) -> None:
    if off_time is None:
        off_time = on_time

    self.set_value(True)
    time.sleep(on_time)
    self.set_value(False)
    time.sleep(off_time)
```

What do you think?

Reply:
Sounds good I'm going to change that